girder/plugin.py

```diff
@@ -0,0 +1,263 @@
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+
+        if not getattr(self, '_loaded', False):
+            # This block is executed on the first call to the function.
+            # The return value of the call (or exception raised) is saved
+            # as attributes on the wrapper for future invocations.
+            self._loaded = True
+            self._exception = None
+            self._return = None
+
+            try:
+                result = func(self, *args, **kwargs)
+            except Exception as e:
+                # If any errors occur in loading the plugin, log the information, and
+                # store failure information that can be queried through the api.
+                self._exception = e
+                logprint.exception('Failed to load plugin %s' % self.name)
+                _pluginFailureInfo[self.name] = {
+                    'traceback': traceback.format_exc()
+                }
+                raise
+
+            _pluginLoadOrder.append(self.name)
+            self._return = result
+            self._success = True
+            logprint.success('Loaded plugin "%s"' % self.name)
+
+        elif self._exception:
+            # If the plugin failed on the first invocation, reraise the original exception.
+            raise self._exception
+
+        return self._return
+
+    return wrapper
+
+
+class _PluginMeta(type):
+    """
+    This is a metaclass applied to the ``GirderPlugin`` descriptor class.  It
+    exists to automatically wrap subclass load methods.
+    """
+    def __new__(meta, classname, bases, classdict):
+        if 'load' in classdict:
+            classdict['load'] = _wrapPluginLoad(classdict['load'])
+        return type.__new__(meta, classname, bases, classdict)
+
+
+@six.add_metaclass(_PluginMeta)
+class GirderPlugin(object):
+    """
+    This is a base class for describing a girder plugin.  A plugin is registered by adding
+    an entrypoint under the namespace ``girder.plugin``.  This entrypoint should return a
+    class derived from this class.
+
+    Example ::
+        class Cats(GirderPlugin):
+
+            def load(self, info):
+                # load dependent plugins
+                girder.plugin.getPlugin('pets').load(info)
+
+                import rest  # register new rest endpoints, etc.
+    """
+    #: This is the named displayed to users on the plugin page.  Unlike the entrypoint name
+    #: used internally, this name can be an arbitrary string.
+    DISPLAY_NAME = None
+
+    #: The name of the web client package on npm.
+    NPM_PACKAGE_NAME = None
+
+    #: The version of the web client package to use.  Accepts any npm version identifier.
+    NPM_PACKAGE_VERSION = '*'
+
+    #: The path of the plugin's web client source code.  This path is given relative to the python
+    #: package.  This property is used to link the web client source into the staging area while
+    #: building in development mode.
+    CLIENT_SOURCE_PATH = 'web_client'
```

Comment:
Could we maybe make `CLIENT_SOURCE_PATH` the primary determinant of whether a package has a web client location? This would have two benefits:
* Allow both `NPM_PACKAGE_NAME` and `NPM_PACKAGE_VERSION` to be hidden and unimportant to most developers.
* Allow the `CLIENT_SOURCE_PATH` path to be more transparent to all developers, by making it explicitly specified and no longer have a magical default.

The result is that developers still need to specify one property (now `CLIENT_SOURCE_PATH` instead of `NPM_PACKAGE_NAME`), but Girder takes care of figuring out `NPM_PACKAGE_NAME` itself from the plugin's `package.json` file.